HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +111,22 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                   # Add locks if they do not belong to the same lock, reduces locking for correlated data
+                   if dest[idx]/N==src[idx]/N:
+                     acquire(&locks[dest[idx]/N])
+                     counts[dest[idx]] += 1
+                     counts[src[idx]] -= 1
+                     release(&locks[src[idx]/N])
```

Comment:
For fine-grained, it is not partitioned and thus checking if dest[idx]/N==src[idx]/N will be redundant.